[AutoDiff] Support differentiation of wrapped properties.

dan-zheng opened the pull request:
Differentiation now correctly supports stored properties with attached property wrappers.

### `Differentiable` derived conformances

`Differentiable` conformance derivation now "peers through" property wrappers.
This behavior is always desirable.

Synthesized `TangentVector` structs contain wrapped properties' `TangentVector`s
as stored properties, not wrappers' `TangentVector`s.

It is irrelevant whether property wrapper types conform to `Differentiable`.
Property wrapper types are required to provide a setter for `var wrappedValue`,
which is needed to synthesize `mutating func move(along:)`.

### Differentiation transform

The differentiation transform is also updated to support differentiation of
wrapped value accessors.

This is done via a new pullback generation code path for "semantic member
accessors". "Semantic member accessors" are accessors attached to member
properties that have a corresponding tangent stored property in the parent
`TangentVector` type.

These accessors have special-case pullback generation based on their semantic
behavior. Currently, only getters and setters are supported.

This special-case pullback generation is currently used for stored property
accessors and property wrapper wrapped value accessors. In the future, it can
also be used to support `@differentiable(useInTangentVector)` computed
properties: SR-12636.

User-defined accesors cannot use this code path because they may use custom
logic that does not semantically perform a member access.

### Example

```swift
import _Differentiation

@propertyWrapper
struct Wrapper<Value> {
  var wrappedValue: Value
}

struct Struct {
  @Wrapper var x: Float = 0
  // Compiler generates:
  // var _x: Wrapper<Float>
  // var x: Float {
  //   get { _x.wrappedValue }
  //   set { _x.wrappedValue = newValue }
  // }
}

extension Struct: Differentiable {
  // Compiler now synthesizes:
  // struct TangentVector: Differentiable & AdditiveArithmetic {
  //   var x: Float
  //   ...
  // }
}

@differentiable
func testWrappedValueAccessors(_ s: Struct, _ x: Float) -> Float {
  var s = s
  // `Struct.x.get` and `Struct.x.set` are both differentiable.
  s.x = s.x * x
  // `modify` accessors are not yet supported, so the line below triggers a diagnostic.
  // s.x *= x
  return s.x
}
```

---

Resolves all of SR-12637 that can be supported today.

dan-zheng committed:
[AutoDiff] Make Differentiable derivation support property wrappers.

Differentiable conformance derivation now "peers through" property wrappers.

Synthesized TangentVector structs contain wrapped properties' TangentVectors as
stored properties, not wrappers' TangentVectors.

Property wrapper types are not required to conform to `Differentiable`.
Property wrapper types are required to provide `wrappedValue.set`, which is
needed to synthesize `mutating func move(along:)`.

```
import _Differentiation

@propertyWrapper
struct Wrapper<Value> {
  var wrappedValue: Value
}

struct Struct: Differentiable {
  @Wrapper var x: Float = 0

  // Compiler now synthesizes:
  // struct TangentVector: Differentiable & AdditiveArithmetic {
  //   var x: Float
  //   ...
  // }
}
```

Resolves SR-12638.

dan-zheng committed:
[AutoDiff] Support property wrapper differentiation.

Support differentiation of property wrapper wrapped value getters and setters.

Create new pullback generation code path for "semantic member accessors".
"Semantic member accessors" are attached to member properties that have a
corresponding tangent stored property in the parent `TangentVector` type.

These accessors have special-case pullback generation based on their semantic
behavior. Currently, only getters and setters are supported.

This special-case pullback generation is currently used for stored property
accessors and property wrapper wrapped value accessors. In the future, it can
also be used to support `@differentiable(useInTangentVector)` computed
properties: SR-12636.

User-defined accesors cannot use this code path because they may use custom
logic that does not semantically perform a member access.

Resolves SR-12639.

dan-zheng committed:
[AutoDiff] Add SR-12642 negative test.

Add SR-12642 negative test: crash regarding `Differentiable` derived
conformances and redeclared properties.

rxwei commented on lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:
```diff
@@ -510,6 +512,232 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
+  if (accessor->getAccessorKind() != AccessorKind::Get &&
+      accessor->getAccessorKind() != AccessorKind::Set)
+    return false;
+  // Currently, only implicit accessors are supported.
```
implicit accessors include getters/setters synthesized for stored properties. I'm not sure checking for implicitness is enough. I think we should check whether `AccessorDecl`'s parent decl is a `VarDecl` that has a property wrapper on it.

dan-zheng replied:
I made the check more explicit in https://github.com/apple/swift/pull/31173/commits/49c4441fc48762261467e7d3893fb6942c051ab1.

I intentionally included stored property accessors to validate the new code path, and because the new code path has more room for micro-optimization. But I'm fine with reverting it.

dan-zheng replied:
Another idea: we can lift the following check into `isSemanticMemberAccessor`: "`TangentVector` must be a struct that contains a stored property with the same name and corresponding `TangentVector` type of the given property".

If we do this, the new code path will be triggered only if it is guaranteed not to emit a non-differentiability error, which may slightly improve diagnostic locations. I'd prefer to investigate this in a follow-up.

rxwei commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2723,13 +2723,22 @@ WARNING(differentiable_nondiff_type_implicit_noderivative_fixit,none,
-      "requires all stored properties to be mutable; use 'var' instead, or add "
-      "an explicit '@noDerivative' attribute"
+      "requires all stored properties not marked with `@noDerivative` to be "
+      "mutable; add '%0.wrappedValue.set', or add an explicit '@noDerivative' "
```
* `.set` is not a user-facing identifier today. I'd suggest describing this in plain English, e.g. "a setter for 'Foo.wrappedValue'".
* Suggesting that users add a setter isn't always helpful. Users are most likely using a property wrapper that they did not define, and adding a setter isn't something that can be done retroactively. I'd suggest removing this suggestion.

dan-zheng replied:
> * I'd suggest removing this suggestion.

Done in https://github.com/apple/swift/pull/31173/commits/f4b389c30b8726977b1ba325cdded9bf0a1c2572.

Alternatively, we could conditionally suggest "adding a setter for 'Wrapper.wrappedValue'" iff the wrapper type is defined in the same file or module.

dan-zheng commented:
Force-pushed to squash minor commits. Testing now.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
There's an unrelated macOS validation test failure, for other PRs too.
```
00:52:56 Failing Tests (1):
00:52:56     Swift(macosx-x86_64) :: api-digester/compare-dump-abi.swift
```

I'll just trigger smoke tests.
@swift-ci Please smoke test macOS

theblixguy commented on lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:
```diff
@@ -510,6 +512,242 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
+  auto *decl = dc->getAsDecl();
+  if (!decl)
+    return false;
+  auto *accessor = dyn_cast<AccessorDecl>(dc->getAsDecl());
```
```suggestion
  auto *accessor = dyn_cast<AccessorDecl>(decl);
```

dan-zheng replied:
Done, thanks.

theblixguy commented on lib/SILOptimizer/Utils/Differentiation/PullbackEmitter.cpp:
```diff
@@ -510,6 +512,242 @@ void PullbackEmitter::addToAdjointBuffer(SILBasicBlock *origBB,
+  // logic that does not semantically perform a member access.
+  // TODO(SR-12636): Support `@differentiable(useInTangentVector)` computed
+  // properties.
+  return false;
```
You could just `return varDecl->hasAttachedPropertyWrapper();` here.

theblixguy commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -39,14 +40,23 @@ getStoredPropertiesForDifferentiation(NominalTypeDecl *nominal, DeclContext *DC,
+          vd->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto *wrappedValueDecl =
+          wrapperDecl->getPropertyWrapperTypeInfo().valueVar;
+      if (!wrappedValueDecl->getAccessor(AccessorKind::Set))
```
I think you can perform this check on `originalProperty` directly - it won’t have a setter if the `wrappedValue` doesn’t have one.

dan-zheng replied:
Done. Thanks for the review!

theblixguy commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -484,20 +494,52 @@ static void addAssociatedTypeAliasDecl(Identifier name, DeclContext *sourceDC,
+                diag::
+                    differentiable_immutable_wrapper_implicit_noderivative_fixit,
+                wrapperDecl->getNameStr(), nominal->getName(),
+                nominalCanDeriveAdditiveArithmetic)
```
The formatting here looks a little strange to me.

dan-zheng replied:
I just use `git clang-format` as an auto-formatter.

dan-zheng committed:
[AutoDiff] Finish wrapped property differentiation support.

Add special-case VJP generation support for "semantic member accessors".
This is necessary to avoid activity analysis related diagnostics and simplifies
generated code.

Fix "wrapped property mutability" check in `Differentiable` derived conformnances.
This resolves SR-12642.

Add e2e test using real world property wrappers (`@Lazy` and `@Clamping`).

dan-zheng commented:
I pushed some final fixes to finish wrapped property differentiation:

> [AutoDiff] Finish wrapped property differentiation support.
> 
> Add special-case VJP generation support for "semantic member accessors".
> This is necessary to avoid activity analysis related diagnostics and simplifies
> generated code.
> 
> Fix "wrapped property mutability" check in `Differentiable` derived conformnances.
> This resolves SR-12642.
> 
> Add e2e test using real world property wrappers.

This PR is ready to land now.

dan-zheng commented:
@swift-ci Please test
